File: CS/RichTextEditorForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.Data.Browsing;

namespace RepCustomXRRichText {
    public partial class RichTextEditorForm : Form {
        public RichTextEditorForm() {
            InitializeComponent();
        }

        private void RichTextEditorForm_Load(object sender, EventArgs e)
        {
            richTextBox1.AllowDrop = true;
            richTextBox1.DragEnter += new DragEventHandler(richTextBox1_DragEnter);
            richTextBox1.DragOver += new DragEventHandler(richTextBox1_DragOver);
            richTextBox1.DragDrop += new DragEventHandler(richTextBox1_DragDrop);
        }

        void richTextBox1_DragEnter(object sender, DragEventArgs e) {
            if(e.Data.GetData(typeof(DataInfo[])) != null)
                e.Effect = DragDropEffects.Copy;
            else
                e.Effect = DragDropEffects.None;
        }

        void richTextBox1_DragOver(object sender, DragEventArgs e) {
            if (e.Data.GetData(typeof(DataInfo[])) != null)
                e.Effect = DragDropEffects.Copy;
            else
                e.Effect = DragDropEffects.None;
        }

        void richTextBox1_DragDrop(object sender, DragEventArgs e) {
            DataInfo droppedData = ((DataInfo[])(e.Data.GetData(typeof(DataInfo[]))))[0];

            richTextBox1.SelectedText = "[" + droppedData.Member + "]";
        }

    }
}

File: VB/RichTextEditorForm.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.Data.Browsing

Namespace RepCustomXRRichText
    Partial Public Class RichTextEditorForm
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub RichTextEditorForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
            richTextBox1.AllowDrop = True
            AddHandler richTextBox1.DragEnter, AddressOf richTextBox1_DragEnter
            AddHandler richTextBox1.DragOver, AddressOf richTextBox1_DragOver
            AddHandler richTextBox1.DragDrop, AddressOf richTextBox1_DragDrop
        End Sub

        Private Sub richTextBox1_DragEnter(ByVal sender As Object, ByVal e As DragEventArgs)
            If e.Data.GetData(GetType(DataInfo())) IsNot Nothing Then
                e.Effect = DragDropEffects.Copy
            Else
                e.Effect = DragDropEffects.None
            End If
        End Sub

        Private Sub richTextBox1_DragOver(ByVal sender As Object, ByVal e As DragEventArgs)
            If e.Data.GetData(GetType(DataInfo())) IsNot Nothing Then
                e.Effect = DragDropEffects.Copy
            Else
                e.Effect = DragDropEffects.None
            End If
        End Sub

        Private Sub richTextBox1_DragDrop(ByVal sender As Object, ByVal e As DragEventArgs)
            Dim droppedData As DataInfo = DirectCast(e.Data.GetData(GetType(DataInfo())), DataInfo())(0)

            richTextBox1.SelectedText = "[" & droppedData.Member & "]"
        End Sub

    End Class
End Namespace